src/Simplex/Chat/Library/Commands.hs

```diff
@@ -2023,75 +2023,171 @@ processChatCommand' vr = \case
+        pure $ partitionEithers mems_
+        where
+          changeRole :: GroupMember -> CM GroupMember
+          changeRole m@GroupMember {groupMemberId, memberContactId, localDisplayName = cName} = do
+            withFastStore (\db -> (,) <$> mapM (getContact db vr user) memberContactId <*> liftIO (getMemberInvitation db user groupMemberId)) >>= \case
+              (Just ct, Just cReq) -> do
+                sendGrpInvitation user ct gInfo (m :: GroupMember) {memberRole = newRole} cReq
+                withFastStore' $ \db -> updateGroupMemberRole db user m newRole
+                pure (m :: GroupMember) {memberRole = newRole}
+              _ -> throwChatError $ CEGroupCantResendInvitation gInfo cName
+      changeRoleCurrentMems :: User -> Group -> [GroupMember] -> CM ([ChatError], [GroupMember], [AChatItem])
+      changeRoleCurrentMems user (Group gInfo members) memsToChange = case L.nonEmpty memsToChange of
+        Nothing -> pure ([], [], [])
+        Just memsToChange' -> do
+          let events = L.map (\GroupMember {memberId} -> XGrpMemRole memberId newRole) memsToChange'
+          (msgs_, _gsr) <- sendGroupMessages user gInfo members events
+          let itemsData = zipWith (fmap . sndItemData) memsToChange (L.toList msgs_)
+          cis_ <- saveSndChatItems user (CDGroupSnd gInfo) itemsData Nothing False
+          when (length cis_ /= length memsToChange) $ logError "changeRoleCurrentMems: memsToChange and cis_ length mismatch"
+          (errs, changed) <- lift $ partitionEithers <$> withStoreBatch' (\db -> map (updMember db) memsToChange)
+          let acis = map (AChatItem SCTGroup SMDSnd (GroupChat gInfo)) $ rights cis_
+          pure (errs, changed, acis)
+          where
+            sndItemData :: GroupMember -> SndMessage -> NewSndChatItemData c
+            sndItemData GroupMember {groupMemberId, memberProfile} msg =
+              let content = CISndGroupEvent $ SGEMemberRole groupMemberId (fromLocalProfile memberProfile) newRole
+                  ts = ciContentTexts content
+                in NewSndChatItemData msg content ts M.empty Nothing Nothing Nothing
+            updMember db m = do
+              updateGroupMemberRole db user m newRole
+              pure (m :: GroupMember) {memberRole = newRole}
+  APIBlockMembersForAll groupId memberIds blockFlag -> withUser $ \user ->
+    withGroupLock "blockForAll" groupId . procCmd $ do
+      Group gInfo members <- withFastStore $ \db -> getGroup db vr user groupId
+      when (selfSelected gInfo) $ throwChatError $ CECommandError "can't block/unblock self"
+      let (blockMems, remainingMems, maxRole, anyAdmin) = selectMembers members
+      when (length blockMems /= length memberIds) $ throwChatError CEGroupMemberNotFound
+      when (length memberIds > 1 && anyAdmin) $ throwChatError $ CECommandError "can't block/unblock multiple members when admins selected"
+      assertUserGroupRole gInfo $ max GRModerator maxRole
```

Comment:
changed check to GRModerator, see comment in deleted code:
`-- TODO GRModerator when most users migrate`